Attachment categories

damianhxy opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Summary
<!-- reviewpad:summarize:start -->
### Summary generated by Reviewpad on 10 Nov 23 20:36 UTC
This pull request includes the following changes:

1. The addition of a new file called "attachment.txt" in the "spec/fixtures/attachments" directory. The content of the file is "Default attachment" and there is no newline at the end of the file.

2. The addition of a new file called "assessment.txt" in the "spec/fixtures/attachments" directory. The content of the file is "Assessment attachment file" and there is no newline at the end of the file.

3. Changes in the `AttachmentsController` related to the `index` action, the `set_attachment` method, and the `attachment_params` method.

4. Changes in the `style.css.scss` file related to the `font-family` style properties.

5. The addition of a new file called `attachments.rb` in the `spec/factories` directory, which defines a factory for the `Attachment` model with various attributes.

6. The addition of a new file called "course.txt" in the "spec/fixtures/attachments" directory. The content of the file is "Course attachment file" and there is no newline at the end of the file.

7. Changes in the `show.html.erb` file related to the display of attachments.

8. Changes in the `assessments_controller.rb` file related to the handling of attachments.

9. A new migration file called "remove_released_from_attachments.rb" that removes the `released` column from the `attachments` table.

10. Changes in the `controller_macros.rb` file related to the creation of attachments using `FactoryBot`.

11. Changes in the `_attachment.html.erb` template file related to the display of attachments.

12. A new migration file called "add_category_name_to_attachments.rb" that adds a `category_name` column to the `attachments` table.

13. Changes in a file related to an attachment form, including JavaScript code and HTML modifications.

14. Changes in the `app/views/assessments/index.html.erb` file related to the display of assessments and attachments.

15. Changes in the `db/schema.rb` file, including an update of the schema version and the addition of columns to the `attachments` table.

16. Changes in the `Attachment` model, including validations, scopes, and methods related to category name, release date, assessments, and file uploads.

17. A migration file named "20230924161219_add_release_at_to_attachments.rb" that adds a `release_at` column to the `attachments` table.

18. A new file called `_course_attachment.html.erb` that contains HTML code for displaying attachments on a course homepage.

Please review these changes and provide your feedback.
<!-- reviewpad:summarize:end -->

## Description
<!--- Describe your changes in detail -->
- Only display course attachments on course landing page (previously: also assessment attachments)
- Add `category_name` and `release_at` fields, remove `released` field
- Group attachments into categories on course landing page, sorted by release date
- Add delete attachment button to edit attachment page
- Backport attachment tests from `develop`

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Improves attachments feature by adding categories (to group related attachments), and release date (to automatically release attachments).

This facilitates the distribution of course material. 122 previously used assessment writeups for this purpose, but this is clunky since it requires multiple clicks to download the files, whereas attachments provide a one-click download functionality.

**Note: unreleased attachments will be set to be released one year in the future**

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

**Note: mostly covered by tests**

Create two attachments under different categories, one unreleased and one released:

**Student view**
![Screenshot 2023-12-07 at 15 02 59](https://github.com/autolab/Autolab/assets/9074856/d7c27d1c-c19d-4597-b8a7-fe5cc195ad61)
Clicking anywhere in the box downloads the attachment

**Instructor view**
![Screenshot 2023-12-07 at 15 02 53](https://github.com/autolab/Autolab/assets/9074856/07920661-8775-4f30-8257-9aa7a4cec721)
Clicking on the name downloads the attachment

- Ensure that assessment attachments do not appear on course landing page
- Ensure that unreleased attachments do not appear to students

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have run rubocop for style check. If you haven't, run `overcommit --install && overcommit --sign` to use pre-commit hook for linting
- [ ] My change requires a change to the documentation, which is located at [Autolab Docs](https://github.com/autolab/docs)
- [ ] I have updated the documentation accordingly, included in this PR

lykimchee commented on app/views/attachments/_form.html.erb:
```diff
@@ -14,13 +18,15 @@
-      <div>
-        <%= f.check_box :released, help_text: "Checking this box will release the attachment to students" %>
-      </div>
+      <%= f.datetime_select :release_at, help_text: "Release attachment after this time" %>
```
Is it possible to add text after this that will let you know whether or not the attachment will be released based on what date the user inputted?

So potentially instead of:
<img width="1061" alt="image" src="https://github.com/autolab/Autolab/assets/66969001/af66eea8-13cb-4311-9b55-0559acc36962">

It would look maybe something like:
(pretend the date has already passed or something so it's going to be released :sob:)
<img width="914" alt="image" src="https://github.com/autolab/Autolab/assets/66969001/1a72e92f-ff21-40a7-a639-190e6b97bae1">

(if assignment will be unreleased)
<img width="554" alt="image" src="https://github.com/autolab/Autolab/assets/66969001/9cd87a21-b115-4da9-8236-4ce2ce4da3ba">

If you can find a better wording for the status message it would be great! But I think since we removed the checkbox it would be nice for users to somehow know, with clarity, whether or not their assessment will be released upon creation!

Also, another small UX thing I noticed: the date shows up as gray.
<img width="224" alt="image" src="https://github.com/autolab/Autolab/assets/66969001/1f06fa11-3c86-427f-bb6f-dfd7d4acf977">

I think this makes sense when the date is filled in with the default value. But when the user changes the date, I think it should be black, like the rest of the input fields?

Or, you could even just do gray for any dates in the past, black for dates in the future (so further letting the user know about the attachment released status upon creation)?

damianhxy replied:
The gray text would be out of the scope for this PR;
The reason why it is grayed is because of materialize's style for readonly fields. It is set to readonly because we are using flatpickr's "altInput" option, which allows us to show one version of a date to the user while internally using another format.

In a sense, this is correct behavior, since the field is indeed readonly. I don't think it's a major deal as of now.

damianhxy replied:
Added code to indicate if attachment will be released immediately / how long in the future
<img width="332" alt="Screenshot 2023-11-10 at 15 17 00" src="https://github.com/autolab/Autolab/assets/9074856/b3c4738c-fe00-4f9e-8db9-393657a3d749">
<img width="280" alt="Screenshot 2023-11-10 at 15 17 08" src="https://github.com/autolab/Autolab/assets/9074856/6956f576-4325-4fcb-852a-232d9ca70f7c">

lykimchee reviewed (changes requested):
Code overall LGTM!

One question I have is: will the category_name only be used for released/unreleased? Or are there other use cases I should be aware about?
If category_name is only used for released/unreleased, then maybe another name for the field would be more intuitive (e.g. released_status, tag, etc.)

### Testing
- [x] Only see course attachments on course landing page
- [x] Only see assessment attachments inside each assessment
- [x] Category (unreleased) shows up if attachment not released
- [x] Creating attachment in both course + assessment
- [x] Deleting attachment from edit attachment page
- [x] Attachment groups
- [x] Attachments sorted by release date

### Frontend Nitpicks
1) <img width="818" alt="Screenshot 2023-10-25 at 2 20 12 PM" src="https://github.com/autolab/Autolab/assets/66969001/5ae4305e-75dd-44de-ac62-f2d37f438dbd">
For some reason, my attachment names always have a bullet point on them?
Also, if the name is too long then it'll shift the edit button down in this weird format

2) <img width="833" alt="Screenshot 2023-10-25 at 2 23 58 PM" src="https://github.com/autolab/Autolab/assets/66969001/3d6c9bec-403f-43ed-9481-789f361e4530">
In the case that the label overflows to the next line, I think it would be neater to have the button left-aligned with the download button instead of what's shown in the screenshot

3) <img width="413" alt="Screenshot 2023-10-25 at 2 33 59 PM" src="https://github.com/autolab/Autolab/assets/66969001/3e4a2152-2072-457c-bf02-60f4a92d52a1">
Inside the edit page the save changes and delete buttons have different stylings

4) <img width="405" alt="image" src="https://github.com/autolab/Autolab/assets/66969001/9fac6da4-8008-4c45-97c9-6b8b04b4562b">
Similarly, upload and save have different styles too.

### UX Considerations

**Unused Category**
In creating an attachment within an assessment, we now have to put in a category, but the category won't show up anywhere:
<img width="837" alt="image" src="https://github.com/autolab/Autolab/assets/66969001/01a51d26-68da-44d9-8bde-a8198f1857c4">
Can we hide the category field when creating an attachment within an assessment?

**Delete Button**
The location of the delete inside the edit attachment page feels like a cancel button (which we should also add, if possible?). Also, the coloring of the delete button makes it seem like a primary action within the page, but the primary action of the edit page should only be saving the changes. I would recommend something like the following:
![image](https://github.com/autolab/Autolab/assets/66969001/01fb0b83-68a2-46cb-a99c-763e8abfde20)
![image](https://github.com/autolab/Autolab/assets/66969001/f501ee36-7e03-46cc-b57a-c6423d93c373)

In general, though, I think having delete be part of the edit functionality isn't very great ([here's](https://ux.stackexchange.com/questions/104751/should-the-delete-action-be-part-of-the-edit-functionality) a stack exchange post, LOL). Which sort of ties in to the next point:

**Clicking on Attachment**
Since the new attachment groups use the same format as the assessment categories, it was weird that I couldn't click into each attachment from the groups (I'm assuming this wasn't a bug?).

In the regular assessment attachments, clicking on the attachment name automatically downloads the file. Could we do the same for the new course attachments, remove the download button -> clicking attachment will automatically download?

Then, you can align the edit/delete buttons to the right in course attachments so it's consistent with assessment attachments, which look like the following:
<img width="825" alt="image" src="https://github.com/autolab/Autolab/assets/66969001/60be4b79-73ae-4a23-88a0-970cb293c65b">

Thanks for reading through this long comment, and let me know if you have questions!! Amazing PR as per always Damian :>>

damianhxy commented on db/migrate/20230924161219_add_release_at_to_attachments.rb:
```diff
@@ -0,0 +1,5 @@
+class AddReleaseAtToAttachments < ActiveRecord::Migration[6.0]
+  def change
+    add_column :attachments, :release_at, :datetime, default: -> { 'CURRENT_TIMESTAMP' }
```
As far as I can tell, the syntax is to use single quotes: https://github.com/rails/rails/issues/27077

damianhxy commented:
> One question I have is: will the category_name only be used for released/unreleased? Or are there other use cases I should be aware about?
If category_name is only used for released/unreleased, then maybe another name for the field would be more intuitive (e.g. released_status, tag, etc.)

category_name is used for the category name displayed for each list, e.g. "General" or "Writing". released/unreleased is determined from `release_at`

> For some reason, my attachment names always have a bullet point on them?
Also, if the name is too long then it'll shift the edit button down in this weird format

Fixed (removed bullet point, ensure that the buttons stick together)
<img width="329" alt="Screenshot 2023-11-05 at 11 27 57" src="https://github.com/autolab/Autolab/assets/9074856/4b763691-ff54-4791-ac13-f353fd0753e7">

damianhxy commented:
> Inside the edit page the save changes and delete buttons have different stylings
Similarly, upload and save have different styles too.

![Screenshot 2023-11-05 at 12 02 39](https://github.com/autolab/Autolab/assets/9074856/74241e9f-fc5a-4f1f-b49b-9d0475a5771d)

Fixed by giving `input` elements the same font-family as everything else (`Source Sans Pro`, previously they were using `sans-serif`)

Also added the same styling for `button` (e.g. problems delete button), `optgroup`, `select` (e.g. version dropdown), `textarea` (e.g. annnotations)

damianhxy commented:
>  Can we hide the category field when creating an attachment within an assessment?

Done (it will automatically default to "General")

damianhxy commented:
> Delete Button
The location of the delete inside the edit attachment page feels like a cancel button (which we should also add, if possible?).

![Screenshot 2023-11-05 at 13 01 20](https://github.com/autolab/Autolab/assets/9074856/a4f64263-c999-4878-8f00-946653716e22)

Removed delete button, added a "cancel" button

>Since the new attachment groups use the same format as the assessment categories, it was weird that I couldn't click into each attachment from the groups (I'm assuming this wasn't a bug?).

This was on purpose, due to the difficulty of having nested links. However, I've updated the logic so that this is how it works for students

![Screenshot 2023-11-05 at 13 03 28](https://github.com/autolab/Autolab/assets/9074856/217b0a03-800c-4fea-b305-6fbd6db70986)

For instructors, I've made the name clickable, removed the download button, aligned the edit/delete buttons to the right

![Screenshot 2023-11-05 at 13 05 49](https://github.com/autolab/Autolab/assets/9074856/c48d4367-ce0f-48fb-8a7a-b49cee57386a)

damianhxy commented on app/models/attachment.rb:
```diff
@@ -6,11 +6,29 @@
+    release_at <= Time.current
+  end
+
   def file=(upload)
     directory = "attachments"
     filename = File.basename(upload.original_filename)
```
Active Storage was implemented in a different commit, will cherrypick in the future

damianhxy commented on app/views/attachments/_form.html.erb:
```diff
@@ -1,10 +1,16 @@
+<% content_for :javascripts do %>
+  <%= javascript_include_tag "initialize_datetimepickers" %>
+<% end %>
```
Leaving it here to avoid conflicting with code on other pages

damianhxy replied:
Will leave this for a future PR, but thanks for the suggestion

damianhxy commented on app/views/attachments/_form.html.erb:
```diff
@@ -14,13 +20,11 @@
     </li>
     <li>
-      <%= f.submit(@attachment.new_record? ? "Create New Attachment" : "Save Changes", class: "btn primary") %>
+      <%= link_to "Cancel", :back, class: "btn-flat" %>
```
There are multiple routes to this page, I want to return the users to the page they came from. Is there a way to fallback if the HTTP_REFERER is not set?

damianhxy replied:
That is too much extra bloat for a button that is meant to be ad hoc. Might pursue in the future if it becomes a problem.

damianhxy commented on app/views/attachments/_form.html.erb:
```diff
@@ -14,13 +20,11 @@
-      <div>
-        <%= f.check_box :released, help_text: "Checking this box will release the attachment to students" %>
-      </div>
+      <%= f.datetime_select :release_at, help_text: "Release attachment after this time" %>
```
How could I implement that?

damianhxy commented on db/migrate/20230924161219_add_release_at_to_attachments.rb:
```diff
@@ -0,0 +1,18 @@
+class AddReleaseAtToAttachments < ActiveRecord::Migration[6.0]
+  def up
+    add_column :attachments, :release_at, :datetime, default: -> { 'CURRENT_TIMESTAMP' }
+    Attachment.find_each do |attachment|
+      if attachment.released
+        attachment.update_attribute(:release_at, DateTime.now)
+      else
+        attachment.update_attribute(:release_at, DateTime.now + 1.year)
+      end
+    end
+  end
+  def down
+    Attachment.find_each do |attachment|
+      attachment.update_attribute(:released, attachment.release_at < DateTime.now)
+    end
+    remove_column :attachments, :release_at
+  end
+end
```
update isn't the same as update_attribute, it takes a hash instead.

Not setting to nil because that would cause more special handling down the road. 1 year is reasonable.

Using released in the down migration should be fine since released should exist (if rolling back the migration after this one, which removed released)

damianhxy commented on db/migrate/20230924161219_add_release_at_to_attachments.rb:
```diff
@@ -0,0 +1,18 @@
+    end
+  end
+  def down
+    Attachment.find_each do |attachment|
+      attachment.update(released: attachment.release_at < DateTime.now)
+    end
```
How would I implement that?

damianhxy commented on db/migrate/20230924161219_add_release_at_to_attachments.rb:
```diff
@@ -0,0 +1,13 @@
+    Attachment.where(released: false).update_all(release_at: DateTime.now + 1.year)
+  end
+  def down
+    Attachment.find_each do |attachment|
+      attachment.update(released: attachment.release_at < DateTime.now)
+    end
```
I will rather use the ORM for this

20wildmanj commented:
I know there were previous discussions about this above, but what are your thoughts about keeping the course card clickable to download but still including the download icon to make it clear that clicking it will result in a download?

![Screenshot 2023-12-04 at 2 35 02 PM](https://github.com/autolab/Autolab/assets/25730111/0ac0697a-51c9-4c12-9e94-452cef3fa5ed)

20wildmanj commented:
Also currently for instructors when looking at all attachments via `/attachments`, all attachments including assessment attachments can be seen, but on the assessments index page, only course attachments can be seen. Do you think this is fine (I wonder if people know the attachments index page even exists tbh)

![Screenshot 2023-12-04 at 2 19 33 PM](https://github.com/autolab/Autolab/assets/25730111/e5b922a6-0874-4172-a2e6-85c3b1ab6de4)

20wildmanj commented:
nit: seems like text and the icons aren't vertically aligned completely
<img width="341" alt="Screenshot 2023-11-28 at 1 28 38 PM" src="https://github.com/autolab/Autolab/assets/25730111/cbf0381f-1353-4af8-b86f-08dc223a0907">

20wildmanj commented:
Nit: can you add more vertical space between the create attachment button and the text about the release date?
![Screenshot 2023-12-04 at 2 18 46 PM](https://github.com/autolab/Autolab/assets/25730111/056d13e0-512a-4840-b5aa-874114f3cbb6)

20wildmanj commented:
![Screenshot 2023-12-04 at 2 46 37 PM](https://github.com/autolab/Autolab/assets/25730111/de256466-23cf-438a-98a7-0e9eeb96cd51)
If you try to create an attachment with the same name as another, the error message does not indicate that the name was the issue

20wildmanj commented:
Functionally LGTM, nice test cases! Some visual things that seem a bit off, but nothing big.

damianhxy commented on db/migrate/20230924161219_add_release_at_to_attachments.rb:
```diff
@@ -0,0 +1,13 @@
+class AddReleaseAtToAttachments < ActiveRecord::Migration[6.0]
+  def up
+    add_column :attachments, :release_at, :datetime, default: -> { 'CURRENT_TIMESTAMP' }
```
This is supported by Rails

damianhxy commented on db/migrate/20230924161219_add_release_at_to_attachments.rb:
```diff
@@ -0,0 +1,13 @@
+    Attachment.where(released: false).update_all(release_at: Time.current + 1.year)
+  end
+  def down
+    Attachment.find_each do |attachment|
+      attachment.update(released: attachment.release_at < Time.current)
```
released is removed in a later migration. Both should be rollbacked together, so released would exist while rolling back this migration.

20wildmanj commented:
Changes LGTM
